master/internal/api_runs.go

```diff
@@ -779,3 +779,137 @@ func archiveUnarchiveAction(ctx context.Context, archive bool, runIDs []int32,
+	runIds []int32, filter *string) (
+	[]*apiv1.RunActionResult, error,
+) {
+	if len(runIds) > 0 && filter != nil {
+		return nil, fmt.Errorf("if filter is provided run id list must be empty")
+	}
+	// Get experiment ids
+	var err error
+	var runCandidates []runCandidateResult
+	getQ := db.Bun().NewSelect().
+		ModelTableExpr("runs AS r").
+		Model(&runCandidates).
+		Column("r.id").
+		ColumnExpr("COALESCE((r.archived OR e.archived OR p.archived OR w.archived), FALSE) AS archived").
+		ColumnExpr("r.experiment_id as exp_id").
+		ColumnExpr("((SELECT COUNT(*) FROM runs r WHERE e.id = r.experiment_id) > 1) as is_multitrial").
+		Join("LEFT JOIN experiments e ON r.experiment_id=e.id").
+		Join("JOIN projects p ON r.project_id = p.id").
+		Join("JOIN workspaces w ON p.workspace_id = w.id").
+		Where("r.project_id = ?", projectID)
+
+	if filter == nil {
+		getQ = getQ.Where("r.id IN (?)", bun.In(runIds))
+	} else {
+		getQ, err = filterRunQuery(getQ, filter)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = getQ.Scan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var results []*apiv1.RunActionResult
+	visibleIDs := set.New[int32]()
+	expIDs := set.New[int32]()
+	expToRun := make(map[int32][]int32)
+	for _, cand := range runCandidates {
+		visibleIDs.Insert(cand.ID)
+		if cand.Archived {
+			results = append(results, &apiv1.RunActionResult{
+				Error: "Run is archived.",
+				Id:    cand.ID,
+			})
+			continue
+		}
+		if cand.IsMultitrial {
+			results = append(results, &apiv1.RunActionResult{
+				Error: fmt.Sprintf("Cannot pause/unpause run '%d' (part of multi-trial).", cand.ID),
+				Id:    cand.ID,
+			})
+			continue
+		}
+		if cand.ExpID == nil {
+			results = append(results, &apiv1.RunActionResult{
+				Error: fmt.Sprintf("Cannot pause run '%d' (no associated experiment).", cand.ID),
+				Id:    cand.ID,
+			})
+			continue
+		}
+		_, ok := expToRun[*cand.ExpID]
+		if ok {
+			expToRun[*cand.ExpID] = append(expToRun[*cand.ExpID], cand.ID)
+		} else {
+			expToRun[*cand.ExpID] = []int32{cand.ID}
+		}
+		expIDs.Insert(*cand.ExpID)
+	}
+	if filter == nil {
+		for _, originalID := range runIds {
+			if !visibleIDs.Contains(originalID) {
+				results = append(results, &apiv1.RunActionResult{
+					Error: fmt.Sprintf("Run with id '%d' not found in project with id '%d'", originalID, projectID),
+					Id:    originalID,
+				})
+			}
+		}
+	}
+	// Pause/Resume experiments
+	var expResults []experiment.ExperimentActionResult
+	var errMsg string
+	if isPause {
+		expResults, err = experiment.PauseExperiments(ctx, projectID, expIDs.ToSlice(), nil)
+		errMsg = "Failed to pause associated experiment: %s"
+	} else {
+		expResults, err = experiment.ActivateExperiments(ctx, projectID, expIDs.ToSlice(), nil)
+		errMsg = "Failed to resume associated experiment: %s"
+	}
+	if err != nil {
+		return nil, err
+	}
+	for _, expRes := range expResults {
+		val, ok := expToRun[expRes.ID]
+		if !ok {
+			val = []int32{-1}
```

Comment:
-1 seems much less useful than a message saying "some invariant in the code was broken and we don't to an action on an experiment we didn't expect to"..

Reply:
Fixed